### ISSUE TITLE
syscall: transparently support new linux io_uring interface

### ISSUE THREAD

**Issue by johanbrandhorst** [golang/go#31908]
A document on the latest linux IO syscall has made the rounds of discussion forums on the internet:

**Comment by ianlancetaylor** [comment]
It should be feasible to fit this approach into our current netpoll framework.  For some programs I think it would reduce the number of threads doing file I/O.  It should potentially reduce the number of system calls required to read from the network.

I'm concerned about coordinating access to the ring.  The approach seems designed for high performance communication between the application and the kernel, but it seems easiest to use for an application that uses a single thread for I/O, or in which each I/O thread uses its own ring.  In Go of course each goroutine is acting independently, and it seems infeasible for each thread to have a separate ring. so that means that goroutines will need to coordinate their access to the I/O ring.  That's fine but on high `GOMAXPROCS` systems I would worry about contention for the ring, a contention that I think doesn't exist in the current `epoll` framework.

**Comment by ianlancetaylor** [comment]
Thinking about this further, it's not clear that it makes sense to use this new interface for network I/O.  It seems that it can only handle a fixed number of concurrent I/O requests, and it's going to be quite hard to make that work transparently and efficiently in Go programs.  Without knowing what the Go program plans to do, we can't allocate the ring appropriately.

If that is true, we would only use it for file I/O, where we can reasonably delay I/O operations when the ring is full.  In that case it would not fit into the netpoll system.  Instead, we might in effect send all file I/O requests to a single goroutine which would add them to the ring, while a second goroutine would sleep until events were ready and wake up the goroutine waiting for them.  That should limit the number of threads we need for file I/O and reduce the number of system calls.

**Comment by CAFxX** [comment]
> Without knowing what the Go program plans to do, we can't allocate the ring appropriately

I was wondering if we could use the map approach:  when it becomes full we allocate a new, bigger one, and start submitting requests to the new one. When the old one doesn't have pending requests anymore we deallocate it.

**Comment by coder543** [comment]
> If that is true, we would only use it for file I/O

File I/O is the main reason I'm interested in `io_uring`. My understanding is that the current async networking stuff works just fine, but making file I/O appear to be interruptible so that goroutines don't block entire OS threads requires using a pool of I/O OS threads. If `io_uring` could remove the need for a pool of OS threads for file I/O, that seems to make it worthwhile.

Whether having a ring per P (as @CAFxX suggests) or just having a single thread dedicated to managing a ring... either solution seems fine. Unless there's some measurable advantage to `io_uring` for networking, I don't think it would be that important to switch the networking code at this point.

**Comment by ianlancetaylor** [comment]
Using a ring per P seems possible, but it means that when a P is idle some M will have to be sleeping in a io_uring getevent call.  And then there will have to be a way to wake up that M if we need it, and some way to avoid confusion if the P gets assigned to a different M.  It may be doable but it seems pretty complicated.

**Comment by JimChengLin** [comment]
I think it is a killer feature of Linux 5.1. `libaio`(a wrapper of Linux kernel native async file io api) is broken or at least flawed, which cannot do async buffered file io and could block unexpectedly. The author of `io_uring` claims all problems that `libaio` has are all solved. It would be greater if Go can utilize the new feature seamlessly.

**Comment by dahankzter** [comment]
Apparently io_uring brings many benefits to networked io as well. Is that not the case or is it just hard to accommodate for the size of the ring?

**Comment by ianlancetaylor** [comment]
@dahankzter See the discussion above.  When I looked at `io_uring` earlier I did not see how to fit it into the Go runtime for network I/O.  I can certainly see how `io_uring` would work well for some network programs, but only ones that know exactly how many concurrent I/O requests they expect to have in flight.  The Go runtime can never know that.  This issue is about transparently using `io_uring`, and I don't see how to do that in Go for network connections.  If you see how to do it, please explain. Thanks.

**Comment by axboe** [comment]
The fixed limit has been removed, and another networking concern was the fixed CQ ring size and the difficulty in sizing that. The latter was fixed with the IORING_FEAT_NODROP support, which guarantees that completion events are never dropped. I think that should take care of most of the concerns?

One thing I've been toying with is the idea of having a shared backend in terms of threads between rings. I think that'd work nicely and would allow you to just setup one ring per thread and not worry too much about overhead.

**Comment by ianlancetaylor** [comment]
@axboe Thanks for the update.  Where is the current API documented?

**Comment by axboe** [comment]
I wrote an update here:

https://kernel.dk/io_uring-whatsnew.pdf

and the man pages in liburing are also up-to-date.

**Comment by spacejam** [comment]
I'm working on [rio](https://github.com/spacejam/rio), a pure-rust misuse-resistant io_uring library, and have been thinking a lot about edge cases that folks are likely to hit. One is exactly the overflow issue @axboe mentions that is addressed with `IORING_FEAT_NODROP`, which I work around in earlier kernel versions by blocking submission on the current number of in-flight requests, and guaranteeing that not more requests than the CQ size are ever in-flight (they could be dropped without that flag if there's a cq overflow). 

In the testing of io_uring with [sled](https://github.com/spacejam/sled) I spin up thousands of database instances, each getting their own uring. This quickly causes ENOMEM to be hit, making this approach infeasible under this kind of test. [details here](https://github.com/spacejam/sled/issues/899). So, you may run into this if you go for a ring-per-goroutine. Maybe ring-per-golang proc.go processor would be OK? Just be aware of this ENOMEM tendency. 

Care needs to be taken when working with linked ops. If a linked SQE fails due to, say, a tcp client disconnecting, it will cancel everything down the chain, even if it wrote some bytes into the buffer passed to io_uring. On newer kernel versions you can use HARDLINK instead of LINK to write the partial data into downstream sockets/files/etc... even when the previous write received an error due to disconnection etc... 

io_uring will change everything. It allows us to dramatically reduce our syscalls (even without SQPOLL mode that spins up a kernel thread, negating the need to do a syscall to submit events). This is vital in a post-meltdown/spectre world where our syscalls have gotten much more expensive. It allows us to do real async file IO without O_DIRECT (but it also works great with O_DIRECT if you're building storage engines etc...). It lets us do things like write broadcast proxies where there's a read followed by a barrier, then many socket writes that read from that same buffer. The goldrush has begun :)

**Comment by coder543** [comment]
@spacejam have you looked into using io_uring for async file I/O? I still find that to be the more compelling use case, since Linux did not have a proper story for async file I/O prior to io_uring, in my limited understanding.

I’m definitely interested in seeing benchmarks of io_uring for networking compared to other async networking approaches, since if io_uring is implemented for file I/O, I would speculate that it might only be an incremental amount of work to support io_uring for network I/O as well.

**Comment by axboe** [comment]
@coder543 It actually started as async file IO, as that only worked for O_DIRECT with Linux. With io_uring, it works for buffered and O_DIRECT.

io_uring also supports networked IO. As per 5.4, you can do the regular read/readv or write/writev on the socket, or you can use send/recvmsg. 5.5 adds connect/accept support, and 5.6 will have support for send/recv as well.

**Comment by spacejam** [comment]
@coder543 For bulk-writing and reading on my 7th gen lenovo x1 carbon with nvme + full disk encryption + ext4 with io_uring (without using registered buffers nor file descriptors nor SQPOLL mode) I can hit 5gbps while writing sequentially with O_DIRECT and 6.5gbps reads while reading sequentially with O_DIRECT. This is the `o_direct` example in the rio repo. While using a threadpool or single thread doing the same reads/writes also with O_DIRECT but using synchronous operations I can scrape 2gbps. dd can hit 4gbps writes for large writes on my system with default settings. I'll be collecting and publishing more statistics as I shift gears from ergonomics around rio to throughput and giving long-tail-latency a haircut for various interesting IO patterns, especially around the observations I'm able to measure around buffer/fd registration for different kinds of workloads.

**Comment by johanbrandhorst** [comment]
High level overview: https://lwn.net/Articles/810414/

**Comment by acln0** [comment]
I think the focus should be on file I/O only, to begin with. Network I/O works just fine using netpoll already. If the model works out for file I/O, we can try it for network I/O also (modulo complications related to deadlines, but those don't seem insurmountable).

A ring per P sounds good to me. Each ring would maintain a cache of off-heap completion tokens which identify the parked goroutine and provide a slot to write the result of the operation to. These tokens would function much like `pollDesc`s function for the existing netpoll implementation, but they would be cached per-P.

To perform I/O, a goroutine acquires the usual locks through `internal/poll`, then enters the runtime. It creates an SQE and submits it to the ring attached to its P, then parks itself. When the completion is handled, the goroutine is woken, much like netpoll currently does.

Some pseudocode:

---

The completion token:
```
//go:notinheap
type uringCompletionToken struct {
	link *uringCompletionToken // links to next token in per-P cache
	cg   guintptr              // the goroutine waiting for the completion
	res  int32                 // written by CQ handler, read by cg
}
```

SQ submission, parking the goroutine doing the submission:
```
	_g_ := getg().m.curg
	_p_ := _g_.m.p

	tok := _p_.uringTokenCache.alloc()
	tok.cg = guintptr(_g_)
	sqe := uringSQE{
		op: _IORING_OP_SOMETHING,
		fd: int32(fd),

		// etc. etc.

		userData: uint64(uintptr(unsafe.Pointer(tok))),
	}
	if errno := uringSubmit(_p_.uring, sqe); errno != 0 {
		// fall back out into regular internal/poll code path
		return 0, 0, notHandled
	}

	gopark(..., ..., waitReasonIOWait, ..., ...)

	// Now that we have woken up, interpret results from the token.
	var (
		n     int
		errno int
	)
	if tok.res < 0 {
		errno = int(-tok.res)
	} else {
		n = int(tok.res)
	}

	// Might have woken on another P: reload, then free the token.
	_p_ = getg().m.p
	_p_.uringTokenCache.free(tok)

	return n, errno, handled
```

Handling completions:
```
	cq := r.nextCQ()
	tok := (*uringCompletionToken)(unsafe.Pointer(uintptr(cq.data)))
	tok.res = cq.res
	goready((*g)(unsafe.Pointer(uintptr(tok.cg))), 4)
```

---

The question of who / what handles completions remains.

If I understand things correctly, after parking the current goroutine, we enter the scheduler and we are executing `findrunnable` with a P, which means that we can check per-P rings rather cheaply. If we have unhandled completions in the local ring, we can add the associated goroutines to the local runqueue and start execution right away.

If `findrunnable`, knowing that there is work pending on the local ring, observes an empty completion ring, it should arrange for something to wait on the ring, so that the P can carry on doing other things. This is the one tricky bit.

I think we can leverage the existing netpoll infrastructure to solve this problem. We associate a non-blocking event file descriptor (as in `eventfd(2)`) to each ring using `io_uring_register(IORING_REGISTER_EVENTFD)`. We register these event file descriptors with netpoll, and we teach netpoll to handle them differently from regular file descriptors: if the M executing netpoll receives a notification on a file descriptor which is part of the uring eventfd set, then it looks for a `*uring` in the epoll user data field rather than a `*pollDesc`, as it currently does. It handles the notification by returning the eventfd to quiescent state, reading completions from associated rings, and pushing the associated goroutines to the global runqueue. If netpoll observes spurious notifications from the eventfds, it should know to ignore them, if polling for completions wasn't requested explicitly for the associated rings.

---

I have insufficient knowledge of the runtime to know whether what I am describing is, in fact, feasible. It is all very hand-wavy, and I haven't attempted an implementation (yet?).

What do you think?

**Comment by agnivade** [comment]
Went with the "single thread for I/O" approach and built a small POC here: https://github.com/agnivade/frodo :) Nothing serious, just something for me to learn about io_uring and do something with it.

There are definitely some hairy edge-cases due to my highly inexperienced knowledge of CGo. But it's at a stage to start playing around. I have thrown in a benchmark just for the sake of it, but it does not mean anything - as it's not an apples-apples comparison, and the CGO boundary alone brings an order of magnitude difference.

**Comment by hodgesds** [comment]
I've also been working on a pure [go library](https://github.com/hodgesds/iouring-go) that initial tests passing for reading from files. I wanted to have somewhat working code before looking at where to integrate it into the runtime. However, I've run into a few things that are rather difficult with Go's memory model regarding memory barriers and dealing with multiple writers to the submit ring, which would be less of an issue in the solution proposed by @acln0.

**Comment by git001** [comment]
@axboe 
> I wrote an update here:
> 
> https://kernel.dk/io_uring-whatsnew.pdf
> 
> and the man pages in liburing are also up-to-date.

The git link point to Facebook workplace, is this intentional?

`https://l.workplace.com/l.php?u=http%3A%2F%2Fgit.kernel.dk%2Fliburing&h=AT3WDhsryImUEhlRMqW221yX-_Eo4Kru9lc1LpqTgxLCOVuCFAptdNw27wNawPNlBzYop81S1n9JiJr_ONyV6TLVtMz-Snbaj3ZU9Woj_R6HQCnYL2makW80B3VORrzUO4GcmxNb`

**Comment by pete-woods** [comment]
> @spacejam do you have any interesting experiences that you can talk about from looking into using io_uring for async file I/O? I still find that to be the more compelling use case, since Linux did not have a proper story for async file I/O prior to io_uring, in my limited understanding.
> 
> I’m definitely interested in seeing benchmarks of io_uring for networking compared to other async networking approaches, since if io_uring is implemented for file I/O, I would speculate that it might only be an incremental amount of work to support io_uring for network I/O as well.

I saw ScyllaDB adopted it, they talked about it in this article: https://www.scylladb.com/2020/05/05/how-io_uring-and-ebpf-will-revolutionize-programming-in-linux/

**Comment by Iceber** [comment]
I'm working on an easy-to-use iouring library for golang (https://github.com/iceber/iouring-go), both file IO and socket IO work fine, but the testing and documentation aren't perfect yet!

**Comment by mvdan** [comment]
@Iceber I don't think it's an option to have the `os` or `net` packages depend on a vendored third party package, though. Unless you were offering to contribute parts of it, like in https://github.com/golang/go/issues/31908#issuecomment-617223327. It seems like the difficulty is more with fitting this into Go's runtime, than implementing the use of io_uring itself.

**Comment by Iceber** [comment]
@mvdan Yes, this is just an experiment for now, and it is optimal for the standard library to provide these features

But really integrating `iouring` into `net` would be a huge and probably difficult change, and I don't really think `net` is going to use iouring!
For file IO you can go for third-party packages such as [iouring-go](https://github.com/Iceber/iouring-go), e.g. in the file server.

I'm also going to try to develop `iouring-net`, which combines iouring and networking to achieve more efficient asynchronous networking with goroutine.
Obviously the third-party package `iouring-net` is not for production, it's just a thought and an experiment.

**Comment by netroby** [comment]
io_uring support native in golang should be very impressive move.

**Comment by godzie44** [comment]
Hi, I am working on a [go-uring](https://github.com/godzie44/go-uring) library. In addition to the liburing port itself, it provides a backend for I/O operations (named reactor). With its help, you can implement net.Listener and net.Conn interfaces, and make a comparison of the reactor (with an io_uring inside) against the standard mechanism - netpoller. [Benchmark](https://github.com/godzie44/go-uring/blob/master/example/echo-server-multi-thread/Benchmark.md) (using the example of an echo server). In addition, on [this benchmark](https://github.com/godzie44/go-uring/blob/master/example/echo-server/benchmark.md), a comparison of the go-uring library against the libring. The results suggest that the ring can at least be an interesting alternative for a non-poller.

**Comment by bohdantrotsenko** [comment]
As of September 2022, can golang runtime detect whether the kernel supports ioring (and switch to using it)?

**Comment by ianlancetaylor** [comment]
No.  This issue is still open.

**Comment by aktau** [comment]
Today I read an experience report on Hacker News where someone describes a way they found to work with io_uring that resulted in actual performance improvements over using epoll(2): https://news.ycombinator.com/item?id=35548968. Might be worth keeping in mind when experimenting with this in the netpoller.

**Comment by beef9999** [comment]
Our coroutine lib uses io_uring to replace epoll, and achieved the truely non-block file I/O.
https://github.com/alibaba/PhotonLibOS

**Comment by pawelgaczynski** [comment]
Hi. I wrote a web framework that is based on io_uring: https://github.com/pawelgaczynski/gain. It is entirely written in Go and achieves really high performance. In a test environment based on a m6i.xlarge AWS EC2 instance machine, kernel 5.15.0-1026-aws and go 1.19, it achieved 490k req/s in the plaintext TechEmpower benchmark. Gain uses the liburing library internal port. It is not a full port. It focuses primarily on the networking part, but I don't see any problem extending the implementation to include the rest of the liburing-supported operations and publishing the port on Github as a standalone package. It would also be worthwhile to create an additional layer of abstraction to use the liburing port in a more idiomatic way, as currently the port is very close to the prototype implemented in C and may not be the most intuitive for Go programmers. If anyone is interested in using my liburing port or would like to help develop it, please contact me by creating an issue in the Gain repository.

**Comment by sxstd001** [comment]
How is the status now?

**Comment by ericvolp12** [comment]
I wasn't sure whether to create a new issue or comment here, but we've run into some significant bottlenecks in the current netpoll implementation on `syscall.EpollWait` when working on very large (192 core) systems that have thousands of TCP sockets making fast requests.

I did a full writeup [here](https://jazco.dev/2024/01/10/golang-and-epoll/) but the tl;dr is that in our use-case, we've got >1,500 TCP connections to ScyllaDB shards and >1,000 client connections talking to our ConnectRPC service. Everything is basically rack-local so the requests are generally sub-millisecond so hundreds if not >1k sockets become ready simultaneously causing a block on `EPollWait` (~65% of CPU usage in our CPU profile) since it's only buffering 128 FDs per call (I'm assuming due to historical max EPoll call kernel support).

We had to solve this by breaking up our binary across many containers on the same host which made the EPollWait completely disappear from our CPU profiles.

**Comment by jakebailey** [comment]
I brought this up in the bluesky thread that the above came from, but one thing to consider is that Docker has now disabled `io_uring` support for security reasons: https://github.com/moby/moby/commit/891241e7e74d4aae6de5f6125574eb994f25e169

> This syncs the seccomp profile with changes made to containerd's default
> profile in [1].
> 
> The original containerd issue and PR mention:
> 
> > Security experts generally believe io_uring to be unsafe. In fact
> > Google ChromeOS and Android have turned it off, plus all Google
> > production servers turn it off. Based on the blog published by Google
> > below it seems like a bunch of vulnerabilities related to io_uring can
> > be exploited to breakout of the container.
> >
> > [2]
> >
> > Other security reaserchers also hold this opinion: see [3] for a
> > blackhat presentation on io_uring exploits.
> 
> For the record, these syscalls were added to the allowlist in [4].
> 
> [1]: containerd/containerd@a48ddf4
> [2]: https://security.googleblog.com/2023/06/learnings-from-kctf-vrps-42-linux.html
> [3]: https://i.blackhat.com/BH-US-23/Presentations/US-23-Lin-bad_io_uring.pdf
> [4]: #39415

This has prompted users of `io_uring` to have to rewrite things back to `epoll` so that their code works in containers (e.g. `bun` https://github.com/oven-sh/bun/pull/7470).

I think it's extremely common for Go code to be running within containers, and if a large bulk of them can't use the feature, that seems pretty unfortunate. Hopefully things get worked out such that `io_uring` isn't getting bulk disabled for being generally unsafe.

**Comment by ianlancetaylor** [comment]
@KyleSanderson See the previous comment.  It appears that `io_uring` is not ready for common cases.

Also, note that for Go it seems likely to help for programs that do a lot of parallel local file I/O.  I don't see why `io_uring` will be a huge gain for the common case of a network server.  I would be happy to hear otherwise.

**Comment by ianlancetaylor** [comment]
Thanks for the information.  This issue is about transparent use of `io_uring` in the Go standard library.  For that case, it definitely matters whether it is supported in the most common case where Go programs are run.

Also, since this is about the standard library, I do think that polling is the relevant consideration here.  I don't yet see how to map the standard library operations onto `io_uring` other than the kind of polling we do today with `epoll`.  While certainly the application could set up groups of system calls, it's not clear to me how the standard library can take the current os and net packages and make that happen transparently.

**Comment by panjf2000** [comment]
While [axboe/liburing/wiki: io_uring and networking in 2023](https://github.com/axboe/liburing/wiki/io_uring-and-networking-in-2023) delivers guidance and best practices of converting networking applications to use `io_uring`, it makes it clear that a change to the I/O model is required. Two key takeaways from that doc: 1) delegate the buffer management to `io_uring`; 2) gather I/O requests and batch them at once, neither of which is part of the current pattern within Go's underlying networking. As Ian said, we still don't see a feasible way that can ***transparently*** support `io_uring` in Go std without any unwanted destruction to the existing APIs. Thus, to support `io_uring` in Go, I'm afraid that we're going to need a new std package that provides a new set of APIs along with reworking the Go runtime tremendously (I speculate).

IMHO, `io_uring` is probably more suited to Go's file I/O at this stage. As for network I/O, like I said above, a new std package might be required, and the trouble is, that it might be Linux-specific. Aside from that, There is still an [ongoing debate](https://github.com/axboe/liburing/issues/189) as to how much of a performance boost can come from replacing `epoll` with `io_uring`. Therefore, I'm not sure it's worth the trouble going down that path, not to mention the fact that we haven't quite tapped into the potential of `epoll` yet, see https://github.com/golang/go/issues/65064#issuecomment-1889140389.

**Comment by CAFxX** [comment]
On the topic of becoming more amenable to transparent support for batching requests, there may be a bit of news cooking:

```
io_uring: support deferring submissions to schedule out time

If IORING_SETUP_SCHED_SUBMIT is enabled for !SQPOLL rings,
io_uring_enter(2) syscalls that submit requests will merely prepare the
selected requests and not go through the submission of them. Submission
will instead be performed when the task that prepared the IO is
scheduled out. This allows for fairly trivial batching of submission
events, even across multiple io_uring_enter(2) syscalls, and for
submission of already prepared IO by the time that the task is scheduled
out anyway.

This is a WIP patch, more of a proof of concept than anything else. The
schedule out handling isn't particularly nice, as submission can
potentially cause a blocking event. If the flow to get there is through
schedule, then we could recurse. This is handled by ensuring that the
pending submission list is fully cleared before any of them are
submitted. Hence if an unlikely recursion event into schedule were to
happen, it will be a no-op as nothing is pending.

Outside of doing lazy batch submissions when the task schedules out
anyway, it also allows more precise control of the full list of
pending requests to issue, as they have all been prepared upfront. This
can be advantageous in situations where better decisions can be made
when the kernel knows exactly what is pending on the submission side.
```

https://x.com/axboe/status/1807862209780806067
https://git.kernel.dk/cgit/linux/commit/?h=io_uring-sched-submit&id=f008f64f52496567bec199a47e9df224d71d963f

**Comment by FZambia** [comment]
Hello,

If we look at CPU profile of an optimized Go network application – in most cases we will see time spent on read/write syscalls. For servers which work with many **different** network connections there is no way to optimize this part currently.

I wonder, may io-uring support by Go runtime help here? It seems that using submission queues may help to batch syscalls to read/write from different network sockets and thus decrease the overhead from system calls? So just having this optimization may give a rather notable effect on performance? I do not see that this should somehow affect existing public Go APIs, just an internal batching algorithm.

**Comment by KyleSanderson** [comment]
> Hello,
> 
> If we look at CPU profile of an optimized Go network application – in most cases we will see time spent on read/write syscalls. For servers which work with many **different** network connections there is no way to optimize this part currently.
> 
> I wonder, may io-uring support by Go runtime help here? It seems that using submission queues may help to batch syscalls to read/write from different network sockets and thus decrease the overhead from system calls? So just having this optimization may give a rather notable effect on performance? I do not see that this should somehow affect existing public Go APIs, just an internal batching algorithm.

Absolutely. This is completely transparent to userspace and is a background thing. The problem is in golang it needs to be done at the scheduler level, the channel synchronization is not tight enough to support this being more performant with those broken abstractions there. I've gotten it pretty damn close even with the insane synchronization overheads, which supports atleast a 4x improvement here. This was on file reads alone.

It is one of those things that is just sitting there waiting to be picked up, same with Windows ioring (even with iocp I've seen positive results), and kqueue on fbsd. Maybe when Google starts looking into the performance area will this become a focus.

Beyond easy pickings here, with no api change needed.

**Comment by nirui** [comment]
Given the activity/updates posted under this issue, I'm not very confident that io_uring support will come out before the release of GTA 6 or Half-Life 3. But I'm still hopeful that it might come out before Fallout 5.

@FZambia

> there is no way to optimize this part currently

In the meantime, you can still try reduce the number of read/write syscalls through buffering (a.k.a putting data to `bufio.Reader`/`bufio.Writer` before reading or sending them in batch), which worked well on TCP pipe.

**Comment by FZambia** [comment]
> In the meantime, you can still try reduce the number of read/write syscalls through buffering (a.k.a putting data to bufio.Reader/bufio.Writer before reading or sending them in batch), which worked well on TCP pipe.

Actually, the reason I came to this issue for io-uring support was not per-socket optimizations—those are already done—but cross-socket optimizations (sending the same data to N sockets with one syscall instead of N).

I recently tried to make a quick proof of concept on top of io_uring to broadcast the same data to many sockets. I was hoping to see reduced application CPU usage, since write syscalls are the top CPU consumer in the application profile under heavy broadcast workloads. Unfortunately, I did not observe any CPU benefit (CPU was slightly worse actually). It seems the overhead of the io_uring batched call is comparable to or even worse at this point than issuing many epoll-based write syscalls. I did not investigate further because I'm not sure I would be able to optimize it further myself. I just wanted to mention that this approach is not the free win for this problem that I initially thought it would be.

**Comment by aktau** [comment]
>  since write syscalls are the top CPU consumer in the application profile under heavy broadcast workloads. Unfortunately, I did not observe any CPU benefit (CPU was slightly worse actually). It seems the overhead of the io_uring batched call is comparable to or even worse at this point than issuing many epoll-based write syscalls

It could be good to run both approaches under linux-perf to be able to observe the kernel stacks as well as the user stacks, and observe why write/send and io_uring have the same overhead.

**Comment by thomas-vilte** [comment]
I've been digging into this issue because I wanted to understand what problem io_uring would actually solve for Go before trying to implement anything in the runtime.

I built a small external Linux/amd64 prototype around regular-file `ReadAt` and compared it with the current blocking path. Most of the work was on an i3-9100F / 4-CPU CachyOS machine, then I reproduced the main mechanisms on an 8-vCPU Ubuntu EC2 host.

The workload is a 512 MiB file, random 1 MiB `ReadAt` on a shared descriptor, 1000 closed-loop clients, and three five-second runs per condition. The bounded-blocking comparison started from a verified nonresident file: zero resident source pages after DONTNEED and immediately before measurement.

Blocking was still considerably faster for resident reads, so I don't have evidence that io_uring is a general file-I/O performance win.

What did stand out was the difference in thread/scheduler behavior.

The Go path I looked at is roughly `os.File.ReadAt -> internal/poll.FD.Pread -> pread`. The M remains occupied while that syscall is executing or blocked, even if its P is retaken. With enough simultaneous slow calls I could get substantial native-thread growth. This isn't one thread per request, and resident reads didn't show the same behavior.

I also tested the simpler alternative on the same revision/workload: limit the number of blocking operations in flight.

That worked better than I expected. Limiting blocking to 16–64 outstanding operations retained roughly 76–83% of unrestricted blocking throughput while reducing median sampled process-task peaks from about 1000 to 21–69.

For comparison, forced-async io_uring with the unbounded io-wq worker limit set to 4 landed in roughly the same throughput range. It used fewer sampled process tasks than blocking at 16 (15 vs 21), but substantially more context switches and slightly worse CPU/op.

I didn't collect queue-inclusive latency for these particular runs, so this is only a throughput/resource-population comparison. A semaphore moves pressure into a queue; it doesn't make that cost disappear.

The case I'm still unsure about is composability. An application can put admission control around its own file I/O, but libraries don't necessarily own a process-wide concurrency policy, and unrelated packages may all be doing file I/O independently. I don't know whether that's something the runtime should solve, but it's the part that makes me hesitant to conclude that a semaphore is always the complete answer.

There were a few useful failures along the way too. The first sustained version of the prototype leaked pinned buffers because ownership could be lost around SQ publication; the small tests hadn't caught it. I also initially thought `io_uring_enter` frequency was a major part of the throughput gap. Reducing enter calls by more than two orders of magnitude barely changed throughput, so that explanation didn't hold up.

`IOSQE_ASYNC` recovered parallelism, but default io-wq settings could themselves grow to roughly 1000 process tasks even with the submitter locked. With a stable issuer, raising the worker limit made throughput plateau while task population kept increasing. I saw the same general pattern on both machines, although the kernels, CPU topology, filesystems and storage were different.

The nonresident-start file becomes resident during execution, so I'm not treating those runs as sustained cold-I/O benchmarks.

I also tried the system-wide `perf` comparison suggested above. Userspace callchains were usable, but this machine has `kptr_restrict=2`. Even with an unstripped `vmlinux` whose build ID matched the running kernel, perf couldn't resolve the recorded kernel frames, so I'm not making a kernel-stack attribution claim from those profiles.

Source, methodology and summary:

https://github.com/thomas-vilte/go-regular-file-io-investigation

Raw runs, exact execution revisions, negative results and limitations:

https://github.com/thomas-vilte/go-regular-file-io-investigation/tree/main/results

At this point I'm mainly interested in two things:

1. Is native-thread pressure from highly concurrent regular-file I/O something the runtime should try to address, or is application/library-level admission control the preferred answer?
2. If this is worth exploring further, what evidence would maintainers want before considering a completion-based design inside Go?

**Comment by ianlancetaylor** [comment]
Thanks for the analysis.

I doubt that it's feasible for the Go runtime to apply a restrictive semaphore to file operations. However, the Go runtime could in principle slow down concurrent file operations. For example, perhaps we could track the average time of file I/O operations. If that time starts to increase, then we could delay new file I/O operations for some period of time in the hopes that that would keep the system from thrashing.

That said, though, I am skeptical that large numbers of a parallel file I/O is a common case for Go programs (as opposed to large numbers of parallel network I/O&mdash;that is definitely a common case). So I'm not sure how much work and complexity is it worth to speed up this case, given that the (presumed) few programs that need it can use special cases instead. It would be interesting to hear about actual programs that do lots of parallel file I/O.

**Comment by thomas-vilte** [comment]
Thanks, that makes sense. I think whether this shows up in real programs is the important question for me to answer next.

One concrete example I found is CockroachDB. It currently has a node-level limit on outstanding SSTable block reads, `storage.block_load.node_max_active`, with a default of 7500. The comment describes it as a last-resort queueing mechanism to avoid memory issues or running into Go's OS-thread limit:

https://github.com/cockroachdb/cockroach/blob/13cb3eb27674b4a981e5c526d04c2387e81efd0b/pkg/storage/pebble.go#L120-L136

Pebble also has an open io_uring prototype issue. They mention that the vast majority of reads are random block reads and that the reading goroutine blocks until the read completes:

https://github.com/cockroachdb/pebble/issues/4684

Their motivation there is mainly reducing CPU usage and exploring `O_DIRECT`, not thread pressure, so I don't want to stretch that into evidence for the same problem. But it does confirm that this kind of concurrent blocking file-read pattern exists in a real Go storage engine.

The CockroachDB example also supports your point that programs which need this can handle it themselves. What I'd like to look for next is whether other Go programs end up implementing similar process-wide limits independently.

Your adaptive throttling idea is interesting too. Given how competitive bounded blocking was in my tests, I think admission/throttling is worth considering separately from io_uring rather than treating completion I/O as the assumed solution.

**Comment by KyleSanderson** [comment]
> Source, methodology and summary:
> 
> https://github.com/thomas-vilte/go-regular-file-io-investigation
>  
> At this point I'm mainly interested in two things:
> 
> 1. Is native-thread pressure from highly concurrent regular-file I/O something the runtime should try to address, or is application/library-level admission control the preferred answer?
> 2. If this is worth exploring further, what evidence would maintainers want before considering a completion-based design inside Go?

Way too much locking in your example. Also you're missing the large benefit here of using IOSQE_IO_LINK, you can open a descriptor, read/write, send data on another already open socket, then close the file descriptor. For filesystems, network appliances and similar that have had to go to rust (where this was implemented what feels like a decade ago now), this would make Go competitive for these applications again.

Once you start doing things like that, you'll see even without runtime participation in scheduling this properly you'll win. You're always going to lose on the unoptimized mpmc queue.

I had to write my own io.Copy (yesterday) because sendfile doesn't seem to be implemented(!), same with LimitReader and SectionReader and walking the tree to the top descriptor. Again, tons of easy wins here for existing and potentially new apps with a callback based design. Similar to how simd is finally coming in (which rust has also had for eons), it'll get better, and it'll take time for opinions to mature.

**Comment by thomas-vilte** [comment]
Thanks, this is useful.

You're right that my prototype is synchronization-heavy. It has a global lock around operation/lifecycle state, a channel feeding one submitter, and per-operation completion signaling. I built it correctness-first, so I wouldn't treat its throughput as the ceiling of what an io_uring design can do.

I also didn't test `IOSQE_IO_LINK`. That's a pretty different case from the independent `ReadAt` operations I measured. Something like open -> read -> send -> close is much more interesting because the dependency chain can stay in the kernel instead of returning to userspace between each step.

Your point that this can still win without runtime scheduler integration is interesting for the question I was asking above. It suggests that some of the larger gains may come from an explicit completion-oriented API rather than transparent support underneath `os.File`/the runtime. I'm not sure where that boundary should be, and I'd still be interested in the maintainers' view on it.

You mentioned earlier in this thread that you had seen at least ~4x on file reads. If you still have that implementation or the workload/results, I'd be interested in reproducing it and comparing the architecture. In particular I'd like to understand what you used instead of the channel/op-table path I have now.

I also checked the `sendfile` point. Current Go does have a sendfile fast path and unwraps `LimitedReader`; the `SectionReader` case is already tracked separately in #61727, so I won't derail this thread into that.

**Comment by ianlancetaylor** [comment]
@KyleSanderson sendfile has been implemented for a while. If there is a case where it wasn't used for `io.Copy` but could have been, please open an issue. Thanks.
